### PR TITLE
[Torii] Index more transaction types

### PR DESCRIPTION
**Pull request by jelilat**
Closes https://github.com/dojoengine/dojo/issues/1422

### CODE REVIEW

**Comment by broody** [comment]
Looking great, I think we could include the type in transaction db table to differentiate between the types.

**Comment by glihm** [comment]
Hey @jelilat!
Any update/feedback you could share to address @broody's comment?
Or @broody should this be addresses in a subsequent PR?

**Comment by jelilat** [comment]
@glihm I'll address it in this PR. On it now

File: crates/torii/migrations/20240207171639_add_txn_type.sql [hunk 1]
```diff
@@ -0,0 +1,4 @@
+ALTER TABLE
+    transactions
+ADD
+    COLUMN transaction_type TEXT NOT NULL;
```
Comment (tarrencev) [inline crates/torii/migrations/20240207171639_add_txn_type.sql:4]: I think we need a default here for backwards compat

Comment (jelilat) [reply crates/torii/migrations/20240207171639_add_txn_type.sql]: You're right. Fixed

**Comment by jelilat** [comment]
@tarrencev @broody Had some time to look at this?

File: crates/torii/core/src/engine.rs [hunk 2]
```diff
@@ -191,26 +214,21 @@ impl<'db, P: Provider + Sync> Engine<'db, P> {
...
+                    .await?;
             }
-        }
 
-        info!("processed block: {}", block.block_number);
-
-        Ok(())
+            Ok(())
+        } else {
+            warn!("transaction receipt not found: {}", transaction_hash);
```
Comment (tarrencev) [inline crates/torii/core/src/engine.rs:229]: In this case, we should probably retry processing this block again

**Review by tarrencev — COMMENTED** [review]
generally lgtm! one suggestion then happy to merge

File: crates/torii/core/src/engine.rs (hunk 2 above)
Comment (jelilat) [reply crates/torii/core/src/engine.rs:229]: No, I don't think that we should outrightly. The line also gets called when the receipt is not of `L1Handler` and `Invoke` txn types. 

I'll update the code to make this distinction and retry for failed fetch

Comment (jelilat) [reply crates/torii/core/src/engine.rs]: Made some changes to retry failed transaction processing

File: crates/torii/core/src/sql.rs [hunk 3]
```diff
@@ -235,19 +235,49 @@ impl Sql {
...
-        let calldata = Argument::String(felts_sql_string(&transaction.calldata));
-        let max_fee = Argument::FieldElement(transaction.max_fee);
-        let signature = Argument::String(felts_sql_string(&transaction.signature));
-        let nonce = Argument::FieldElement(transaction.nonce);
+
+        let transaction_type = match transaction {
+            Transaction::Invoke(_) => "INVOKE",
+            Transaction::L1Handler(_) => "L1_HANDLER",
+            _ => return,
+        };
```
Comment (glihm) [inline crates/torii/core/src/sql.rs:245]: We may have this returned from the match below. But definitely not required for now and may be addressed on subsequent PR to support other transactions. :)

**Review by glihm — APPROVED** [review]
Great work @jelilat, good base to support new transaction types with next network upgrades.

We should also add test in the future on the different transaction type. :+1:

**Comment by jelilat** [comment]
> Great work @jelilat, good base to support new transaction types with next network upgrades.
> 
> We should also add test in the future on the different transaction type. 👍

Thanks @glihm

**Pull request merged**